### PR TITLE
Fix a couple of crash-on-invalid situations with local variables defined inside result builders  [5.4]

### PR DESCRIPTION
**Pull request by slavapestov**
5.4 cherry-pick of https://github.com/apple/swift/pull/35812.

Fixes rdar://problem/73545981.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by Bo886** [comment]
n

**Review by Bo886 — APPROVED** [review]
n

**Review by Bo886 — APPROVED** [review]

File: lib/Sema/BuilderTransform.cpp [hunk 1]
```diff
@@ -268,14 +268,30 @@ class BuilderClosureVisitor
   }
 
   void visitPatternBindingDecl(PatternBindingDecl *patternBinding) {
-    // If any of the entries lacks an initializer, don't handle this node.
-    if (!llvm::all_of(range(patternBinding->getNumPatternEntries()),
-                      [&](unsigned index) {
-            return patternBinding->isExplicitlyInitialized(index);
```
Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:274]: -

**Review by Bo886 — APPROVED** [review]

File: lib/Sema/BuilderTransform.cpp [hunk 2]
```diff
@@ -268,14 +268,30 @@ class BuilderClosureVisitor
   }
 
   void visitPatternBindingDecl(PatternBindingDecl *patternBinding) {
-    // If any of the entries lacks an initializer, don't handle this node.
```
Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:271]: -

**Review by Bo886 — APPROVED** [review]

File: lib/Sema/BuilderTransform.cpp (hunk 2 above)
Comment (Bo886) [reply lib/Sema/BuilderTransform.cpp:271]: -

File: lib/Sema/BuilderTransform.cpp [hunk 3]
```diff
@@ -268,14 +268,30 @@ class BuilderClosureVisitor
...
-    if (!llvm::all_of(range(patternBinding->getNumPatternEntries()),
-                      [&](unsigned index) {
-            return patternBinding->isExplicitlyInitialized(index);
-        })) {
-      if (!unhandledNode)
-        unhandledNode = patternBinding;
```
Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:277]: -

File: lib/Sema/BuilderTransform.cpp [hunk 4]
```diff
@@ -268,14 +268,30 @@ class BuilderClosureVisitor
   }
 
   void visitPatternBindingDecl(PatternBindingDecl *patternBinding) {
-    // If any of the entries lacks an initializer, don't handle this node.
-    if (!llvm::all_of(range(patternBinding->getNumPatternEntries()),
```
Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:272]: -

Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:272]: -

File: lib/Sema/BuilderTransform.cpp (hunk 2 above)
Comment (Bo886) [inline lib/Sema/BuilderTransform.cpp:271]: -

**Review by Bo886 — APPROVED** [review]

**Review by Bo886 — APPROVED** [review]

**Review by Bo886 — APPROVED** [review]

**Review by Bo886 — APPROVED** [review]

**Review by Bo886 — CHANGES_REQUESTED** [review]
n

**Review by Bo886 — APPROVED** [review]

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Review by tkremenek — APPROVED** [review]

**Pull request merged**